scala/scala.bzl

```diff
@@ -21,360 +21,375 @@ _srcjar_filetype = FileType([".srcjar"])
-    ctx.files._jdk + 
-    ctx.files._jar + 
-    ctx.files._scalasdk + 
-    [ctx.outputs.manifest, 
-      ctx.file._ijar, 
-      ctx.file._scalac, 
-      ctx.file._java, 
-      scalac_args_file])
-  if compile_java_srcs:
-    ins.extend([javac_args_file])
-  ctx.action(
-      inputs=ins,
-      outputs=outs,
-      command=cmd,
-      mnemonic="Scalac",
-      progress_message="scala %s" % ctx.label,
-      arguments=[])
 
-def _compile_or_empty(ctx, jars, srcjars, buildijar):
-  # We assume that if a srcjar is present, it is not empty
-  if len(ctx.files.srcs) + len(srcjars) == 0:
-    _build_nosrc_jar(ctx, buildijar)
-    #  no need to build ijar when empty
-    return struct(ijar=ctx.outputs.jar, class_jar=ctx.outputs.jar)
-  else:
-    _compile(ctx, jars, srcjars, buildijar)
-    ijar = None
+    scalac_args = """
+JarOutput: {out}
+Manifest: {manifest}
+ScalacOpts: {scala_opts}
+Plugins: {plugin_arg}
+Classpath: {cp}
+Files: {files}
+EnableIjar: {enableijar}
+ijarOutput: {ijar_out}
+ijarCmdPath: {ijar_cmd_path}
+SourceJars: {srcjars}
+JavacPath: {javac_path}
+JavacOpts: {javac_opts}
+JavaFiles: {java_files}
+ResourceSrcs: {resource_src}
+ResourceDests: {resource_dest}
+""".format(
+        out=ctx.outputs.jar.path,  # 0
+        manifest=ctx.outputs.manifest.path,  # 1
+        scala_opts=",".join(ctx.attr.scalacopts),  # 2
+        plugin_arg=plugin_arg,  # 3
+        cp=compiler_classpath,  # 4
+        files=",".join([f.path for f in sources]),  # 5
+        enableijar=buildijar,
+        ijar_out=ijar_output_path,
+        ijar_cmd_path=ijar_cmd_path,
+        srcjars=",".join([f.path for f in all_srcjars]),
+        javac_opts=" ".join(ctx.attr.javacopts),
+        javac_path=ctx.file._javac.path,
+        java_files=",".join([f.path for f in java_srcs]),
+        resource_src=",".join([f.path for f in ctx.files.resources]),
+        resource_dest=",".join(
+          [_adjust_resources_path(f.path)[1] for f in ctx.files.resources]
+          ),
+        )
+    argfile = ctx.new_file(
+      ctx.outputs.jar,
+      "%s_worker_input" % ctx.label.name
+    )
+    ctx.file_action(output=argfile, content=scalac_args)
+
+    outs = [ctx.outputs.jar]
     if buildijar:
-      ijar = ctx.outputs.ijar
+        outs.extend([ctx.outputs.ijar])
+    ins = (list(jars) +
+           list(dep_srcjars) +
+           list(srcjars) +
+           list(sources) +
+           ctx.files.srcs +
+           ctx.files.plugins +
+           ctx.files.resources +
+           ctx.files._jdk +
+           ctx.files._scalasdk +
+           [ctx.outputs.manifest,
+            ctx.file._ijar,
+            ctx.file._java,
+            argfile])
+    ctx.action(
+        inputs=ins,
+        outputs=outs,
+        executable=ctx.executable._scalac,
+        mnemonic="Scalac",
+        progress_message="scala %s" % ctx.label,
+        execution_requirements={"supports-workers": "1"},
+        arguments=list(ctx.attr.jvm_flags) + ["@" + argfile.path],
```

Comment:
can we add a comment here like:
```
when we run with a worker, the `@argfile.path` is removed and passed line by line as arguments in the protobuf. In that case, the rest of the arguments are passed to the process that starts up and stays resident.

In either case (worker or not), they will be jvm flags which will be correctly handled since the executable is a jvm app that will consume the flags on startup.
```
This confused me a bit, and I may forget why it works.